File: lab3-java/src/main/SharedArray.java
package main;

import java.util.Arrays;
import java.util.concurrent.locks.ReentrantLock;

public class SharedArray {
    private final int[] data;
    private final ReentrantLock lock = new ReentrantLock();

    public SharedArray(int size) {
        this.data = new int[size];
    }

    public Result tryMark(int index, int id) {
        lock.lock();
        try {
            if (index < 0 || index >= data.length) {
                return new Result(false, countById(id));
            }
            if (data[index] == 0) {
                data[index] = id;
                return new Result(true, countById(id));
            }
            return new Result(false, countById(id));
        } finally {
            lock.unlock();
        }
    }

    public void clearById(int id) {
        lock.lock();
        try {
            for (int i = 0; i < data.length; i++) {
                if (data[i] == id) {
                    data[i] = 0;
                }
            }
        } finally {
            lock.unlock();
        }
    }

    public int countById(int id) {
        lock.lock();
        try {
            int count = 0;
            for (int v : data) {
                if (v == id) count++;
            }
            return count;
        } finally {
            lock.unlock();
        }
    }

    public int[] snapshot() {
        lock.lock();
        try {
            return Arrays.copyOf(data, data.length);
        } finally {
            lock.unlock();
        }
    }

    public int size() {
        return data.length;
    }
}

File: lab3-C++/src/shared_array.cpp
#include "shared_array.h"

SharedArray::SharedArray(int size) : data_(size, 0) {}

bool SharedArray::try_mark(int index, int id) {
    std::lock_guard<std::mutex> lock(mtx_);
    if (data_[index] == 0) {
        data_[index] = id;
        return true;
    }
    return false;
}

void SharedArray::clear_by_id(int id) {
    std::lock_guard<std::mutex> lock(mtx_);
    for (int& v : data_) {
        if (v == id) {
            v = 0;
        }
    }
}

int SharedArray::count_by_id(int id) const {
    std::lock_guard<std::mutex> lock(mtx_);
    int c = 0;
    for (int v : data_) {
        if (v == id) {
            ++c;
        }
    }
    return c;
}

int SharedArray::get(int index) const {
    std::lock_guard<std::mutex> lock(mtx_);
    return data_[index];
}

std::vector<int> SharedArray::snapshot() const {
    std::lock_guard<std::mutex> lock(mtx_);
    return data_;
}

int SharedArray::size() const {
    return data_.size();
}